src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -0,0 +1,235 @@
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  buffer_state_[object_id].references--;
+  if (buffer_state_[object_id].references == 0) {
+    chunk_info_.erase(object_id);
+    std::shared_ptr<plasma::PlasmaClient> store_client = buffer_state_[object_id].client;
+    ARROW_CHECK_OK(store_client->Release(ObjectID(object_id).to_plasma_id()));
+    ReleaseObjectStore(store_client);
+    buffer_state_.erase(object_id);
+  }
+  return ray::Status::OK();
+}
+
+ray::Status ObjectBufferPool::AbortGet(const ObjectID &object_id) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  chunk_info_.erase(object_id);
+  std::shared_ptr<plasma::PlasmaClient> store_client = buffer_state_[object_id].client;
+  ARROW_CHECK_OK(store_client->Release(ObjectID(object_id).to_plasma_id()));
+  ReleaseObjectStore(store_client);
+  buffer_state_.erase(object_id);
+  return ray::Status::OK();
+}
+
+const ObjectBufferPool::ChunkInfo &ObjectBufferPool::CreateChunk(
+    const ObjectID &object_id, uint64_t data_size, uint64_t metadata_size,
+    uint64_t chunk_index) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  RAY_LOG(DEBUG) << "CreateChunk " << object_id << " " << data_size << " "
+                 << metadata_size;
+  if (chunk_info_.count(object_id) == 0) {
+    const plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
+    int64_t object_size = data_size - metadata_size;
+    // Try to create shared buffer.
+    std::shared_ptr<Buffer> data;
+    std::shared_ptr<plasma::PlasmaClient> store_client = GetObjectStore();
+    arrow::Status s =
+        store_client->Create(plasma_id, object_size, NULL, metadata_size, &data);
+    std::vector<boost::asio::mutable_buffer> buffer;
+    uint8_t *mutable_data;
+    if (s.ok()) {
+      // Read object into store.
+      mutable_data = data->mutable_data();
+    } else {
+      RAY_LOG(ERROR) << "Buffer Create Failed: " << s.message();
+      // Read object into empty buffer.
+      std::vector<uint8_t> mutable_vec;
+      mutable_vec.resize(data_size);
+      mutable_data = mutable_vec.data();
+      create_failure_buffers_[object_id] = mutable_vec;
+    }
+    uint64_t num_chunks = BuildChunks(object_id, mutable_data, data_size, metadata_size);
+    buffer_state_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(object_id),
+        std::forward_as_tuple(store_client, num_chunks, BufferStateType::CREATE));
+  }
+  RAY_CHECK(buffer_state_[object_id].state_type == BufferStateType::CREATE);
+  return chunk_info_[object_id][chunk_index];
+}
+
+ray::Status ObjectBufferPool::SealOrAbortBuffer(const ObjectID &object_id,
+                                                bool succeeded) {
+  std::lock_guard<std::mutex> lock(pool_mutex_);
+  buffer_state_[object_id].references--;
+  RAY_LOG(DEBUG) << "SealOrAbortBuffer " << object_id << " "
+                 << buffer_state_[object_id].references;
+  if (!succeeded) {
+    buffer_state_[object_id].one_failed = true;
```

Comment:
Thinking about it now, it might make sense to only abort the chunk that failed, but keep any work that was done on the other chunks.